### DISCUSSION TITLE
Clearing singletons

### DISCUSSION THREAD

**Discussion by raysalemi** [pyuvm/pyuvm#64] [Ideas]
It turns out the real problem is that one cannot clear the factory in run_test() because the test class is not in there to be created.  The factory may need to be an independent singleton.

**Comment by jonpovey** [comment]
I'm not sure I have a clear idea what the semantics should be - I just found it surprising that the ConfigDB contained config from a previous test. That doesn't seem right. Seems like the same principle of least surprise should apply to the factory, though I haven't really used the factory yet.
Could things be cleared at the **end** of a test rather than at the beginning? That would still leave the door open to configuration in the `@cocotb.test()` before `run_test()`.

I am not sure how to arrange it, but I increasingly get the feeling that the test writer should not be writing the `@cocotb.test()` function entry point when using pyuvm. That should be abstracted away / auto generated inside pyuvm, some function inside pyuvm would be the first thing to get run, where it could reset singletons or whatever, and the first user code would be a function inside the test class. I am not sure how that should work with parameterising a test to be run multiple times. It's just a feeling so far...

**Reply by raysalemi** [reply]
> 
> Could things be cleared at the **end** of a test rather than at the beginning? That would still leave the door open to configuration in the `@cocotb.test()` before `run_test()`.
> 
> 
> 
That's a great idea!  I'll take a look at it tomorrow.

**Reply by raysalemi** [reply]
The problem is the factory must persist as all the classes get defined once before the tests run. 

I think living within the cocotb test structure is the right approach. We just need to figure out the right clearing mechanism between tests.

**Reply by raysalemi** [reply]
I think I'm going g to create a `clear_singletons` class method that clears every but `uvm_factory`

**Reply by jonpovey** [reply]
I haven't looked closely at the factory, I didn't realise that "classes defined at module loading time" were mixed in with what I imagine as "overrides set later by user code".
I wonder if the factory could include such a split, keeping separate the module-load-time definitions and things-that-happen-later; and throwing away the things set up per test, at the end of a test.

**Reply by raysalemi** [reply]
> I'm not sure I have a clear idea what the semantics should be - I just found it surprising that the ConfigDB contained config from a previous test. That doesn't seem right. Seems like the same principle of least surprise should apply to the factory, though I haven't really used the factory yet. Could things be cleared at the **end** of a test rather than at the beginning? That would still leave the door open to configuration in the `@cocotb.test()` before `run_test()`.

I've resolved this by clearing all the singletons except the factory ones. Then I clear the overrides in those.  This returns the test to a pristine state with the class definitions in place.

**Reply by raysalemi** [reply]
> I am not sure how to arrange it, but I increasingly get the feeling that the test writer should not be writing the `@cocotb.test()` function entry point when using pyuvm. That should be abstracted away / auto generated inside pyuvm, some function inside pyuvm would be the first thing to get run, where it could reset singletons or whatever, and the first user code would be a function inside the test class. I am not sure how that should work with parameterising a test to be run multiple times. It's just a feeling so far...

I wouldn't want to do this as it could set up a feeling of competition between **cocotb** and **pyuvm**, which is definitely not what I want. The big problem has been persistent data, which is now removed.

**Comment by raysalemi** [comment]
I've merged in the "clear_singletons" functionality.  It may break some users, but I'll bet not many.
It now clears all the singletons but the factory.  It keeps the factory but clears the overrides.
You can have old behavior with `run_test("TESTNAME", keep_singletons=True)`
